### PR TITLE
[SFT] Log mean token accuracy from Liger kernel

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

We log the mean token accuracy from Liger kernel if it is available

Requires https://github.com/linkedin/Liger-Kernel/pull/910

to test do:
```bash
pip install git+https://github.com/kashif/Liger-Kernel.git@mean_token_accuracy
pip install git+https://github.com/huggingface/trl.git@iiger-mean-acc
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -1133,8 +1138,12 @@ def compute_loss(
             self._total_train_tokens += num_tokens_in_batch
         self._metrics[mode]["num_tokens"] = [self._total_train_tokens]
 
-        # Compute token accuracy if we have labels and if the model is not using Liger (no logits)
-        if not self.args.use_liger_kernel:
+        if self.args.use_liger_kernel:
+            if hasattr(outputs, "token_accuracy") and outputs.token_accuracy is not None:
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py]: I prefer using an explicit condition, like this:

```python
if self.args.use_liger_kernel and version("liger_kernel") >= "0.6.4":
    token_accuracy = self.accelerator.gather_for_metrics(outputs.token_accuracy).mean().item()
    self._metrics[mode]["mean_token_accuracy"].append(token_accuracy)
```

Having the explicit condition makes it much easier to spot when the check can be removed after we bump the minimum `liger_kernel` version. It’s clearer and more maintainable than hiding the logic elsewhere.

Comment (qgallouedec) [reply trl/trainer/sft_trainer.py]: the question is then, which version will include this feature

Comment (kashif) [reply trl/trainer/sft_trainer.py]: the issue with this is that I cannot test it since the dev version is the current version in their setup

Comment (qgallouedec) [reply trl/trainer/sft_trainer.py]: ah true

**Comment by lewtun** [comment]
Tested with both DDP and FSDP2 - accuracies and loss are ~identical both with and without Liger. 

I suggest we wait until the Liger PR is merged before merging this one with @qgallouedec's change:

Screenshot 2025-11-04 at 20 08 58
Screenshot 2025-11-04 at 20 09 02

**Comment by kashif** [comment]
Liger PR merged... waiting for a release, and then can pin the version to that

File: trl/trainer/sft_trainer.py (hunk 1 above)
Comment (kashif) [reply trl/trainer/sft_trainer.py]: checking explicitly now

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -1151,8 +1158,12 @@ def compute_loss(
...
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py]: Can we have an explicit version check instead of hasattr?

Comment (qgallouedec) [reply trl/trainer/sft_trainer.py]: Also, if you have the right version, there is no reason for token_accuracy to be None, is there?

**Review by qgallouedec — APPROVED** [review]
Lgtm, just a small remark

File: trl/trainer/sft_trainer.py (hunk 2 above)
Comment (kashif) [reply trl/trainer/sft_trainer.py]: there could be, as the outputs depend on how the model is patched... a model might not be patched properly, and then we get an exception for that model. I am fixing models as i find them but not all might have this.
